controllers/provisioning_controller.go

```diff
@@ -152,32 +209,8 @@ func (r *ProvisioningReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
 		return ctrl.Result{}, nil
 	}
 
-	// examine DeletionTimestamp to determine if object is under deletion
-	if baremetalConfig.ObjectMeta.DeletionTimestamp.IsZero() {
-		// The object is not being deleted, so if it does not have our finalizer,
-		// then lets add the finalizer and update the object. This is equivalent
-		// registering our finalizer.
-		if !containsString(baremetalConfig.ObjectMeta.Finalizers,
-			metal3iov1alpha1.ProvisioningFinalizer) {
-			// Add finalizer becasue it doesn't already exist
-			baremetalConfig.ObjectMeta.Finalizers = append(baremetalConfig.ObjectMeta.Finalizers,
-				metal3iov1alpha1.ProvisioningFinalizer)
-			if err := r.Client.Update(context.Background(), baremetalConfig); err != nil {
-				return ctrl.Result{}, errors.Wrap(err, "failed to update Provisioning CR with its finalizer")
-			}
-		}
-	} else {
-		// The Provisioning object is being deleted
-		if containsString(baremetalConfig.ObjectMeta.Finalizers, metal3iov1alpha1.ProvisioningFinalizer) {
-			// Add any specific deletion logic here
-
-			// Remove our finalizer from the list and update it.
-			baremetalConfig.ObjectMeta.Finalizers = removeString(baremetalConfig.ObjectMeta.Finalizers,
-				metal3iov1alpha1.ProvisioningFinalizer)
-			if err := r.Client.Update(context.Background(), baremetalConfig); err != nil {
-				return ctrl.Result{}, errors.Wrap(err, "failed to remove finalizer from Provisioning CR")
-			}
-		}
+	if err = r.checkForCRDeletion(baremetalConfig); err != nil {
```

Comment:
if you have just deleted everything, is it worth continuing?
maybe
```
deleted, err = r.checkForCRDeletion()
if err != nil || deleted {
    return ctrl.Result{}, err
}
```

Reply:
I'd go so far as to say it's essential not to continue, otherwise it'll all get recreated again :)

Reply:
Done.

Reply:
if it has successfully deleted all resouces it will return nil here and then it will re-create the deployment below